File: Sources/custom_adapter/TopOnAdapter/Objc/6.5.x/RewardVideo/AlxTopOnRewardVideoEvent.m
//
//  AlxTopOnRewardVideoEvent.m
//  AlxAdsOCDemo
//

#import "AlxTopOnRewardVideoEvent.h"
#import "AlxTopOnBaseManager.h"
#import "AlxTopOnBiddingRequestManager.h"

static NSString *const TAG = @"AlxTopOnRewardVideoEvent";

@implementation AlxTopOnRewardVideoEvent

- (void)rewardVideoAdLoad:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdLoad", TAG);
    if (self.isC2SBiding) {
        [AlxTopOnBiddingRequestManager disposeLoadSuccessWithPrice:[ad getPrice]
                                                            unitID:self.serverInfo[[AlxTopOnBaseManager unitID]]];
        self.isC2SBiding = NO;
    } else {
        [self trackRewardedVideoAdLoaded:ad adExtra:nil];
    }
}

- (void)rewardVideoAdFailToLoad:(AlxRewardVideoAd *)ad didFailWithError:(NSError *)error {
    NSLog(@"%@: rewardVideoAdFailToLoad: %@", TAG, error.localizedDescription);
    if (self.isC2SBiding) {
        [AlxTopOnBiddingRequestManager disposeLoadFailWithError:error
                                                         unitID:self.serverInfo[[AlxTopOnBaseManager unitID]]];
    } else {
        [self trackRewardedVideoAdLoadFailed:error];
    }
}

- (void)rewardVideoAdImpression:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdImpression", TAG);
    [self trackRewardedVideoAdShow];
}

- (void)rewardVideoAdClick:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdClick", TAG);
    [self trackRewardedVideoAdClick];
}

- (void)rewardVideoAdClose:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdClose", TAG);
    [self trackRewardedVideoAdCloseRewarded:YES extra:@{@"":@""}];
}

- (void)rewardVideoAdPlayStart:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdPlayStart", TAG);
    [self trackRewardedVideoAdVideoStart];
}

- (void)rewardVideoAdPlayEnd:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdPlayEnd", TAG);
    [self trackRewardedVideoAdVideoEnd];
}

- (void)rewardVideoAdReward:(AlxRewardVideoAd *)ad {
    NSLog(@"%@: rewardVideoAdReward", TAG);
    [self trackRewardedVideoAdRewarded];
}

- (void)rewardVideoAdPlayFail:(AlxRewardVideoAd *)ad didFailWithError:(NSError *)error {
    NSLog(@"%@: rewardVideoAdPlayFail", TAG);
//    [self trackRewardedVideoAdPlayFailed:error];
    [self trackRewardedVideoAdPlayEventWithError:error extra:@{@"":@""}];
}

- (NSString *)networkUnitId {
    return self.serverInfo[[AlxTopOnBaseManager unitID]] ?: @"";
}

- (void)dealloc {
    NSLog(@"%@: dealloc", TAG);
}

@end

File: AlxAdsDemo/TopOnAdapter/AlxTopOnRewardVideoEvent.swift
//
//  AlxTopOnRewardVideoEvent.swift
//

import Foundation
import AnyThinkSDK
import AnyThinkRewardedVideo
import AlxAds

@objc(AlxTopOnRewardVideoEvent)
public class AlxTopOnRewardVideoEvent: ATRewardedVideoCustomEvent,AlxRewardVideoAdDelegate {
    
    private static let TAG = "AlxTopOnRewardVideoEvent"
    
    private var isRewarded:Bool = false
    
    public override init(info serverInfo: [AnyHashable : Any], localInfo: [AnyHashable : Any]) {
        super.init(info: serverInfo, localInfo: localInfo)
    }
    
    public func rewardVideoAdLoad(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdLoad",AlxTopOnRewardVideoEvent.TAG)
        if self.isC2SBiding {
            AlxTopOnBiddingRequestManager.disposeLoadSuccess(price: ad.getPrice(), unitID: self.serverInfo[AlxTopOnBaseManager.unitID] as? String)
            self.isC2SBiding = false
        }else{
            self.trackRewardedVideoAdLoaded(ad, adExtra: nil)
        }
    }
    
    public func rewardVideoAdFailToLoad(_ ad: AlxRewardVideoAd, didFailWithError error: Error) {
        NSLog("%@: rewardVideoAdFailToLoad",AlxTopOnRewardVideoEvent.TAG)
        if self.isC2SBiding {
            AlxTopOnBiddingRequestManager.disposeLoadFail(error: error, unitID: self.serverInfo[AlxTopOnBaseManager.unitID] as? String)
        }else{
            self.trackRewardedVideoAdLoadFailed(error)
        }
    }
    
    public func rewardVideoAdImpression(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdImpression",AlxTopOnRewardVideoEvent.TAG)
        self.trackRewardedVideoAdShow()
    }
    
    public func rewardVideoAdClick(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdClick",AlxTopOnRewardVideoEvent.TAG)
        self.trackRewardedVideoAdClick()
    }
    
    public func rewardVideoAdClose(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdClose",AlxTopOnRewardVideoEvent.TAG)
        let closeType:ATAdCloseType = .unknow
        self.trackRewardedVideoAdCloseRewarded(isRewarded, extra: [kATADDelegateExtraDismissTypeKey:closeType])
    }
    
    public func rewardVideoAdPlayStart(_ ad:AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdPlayStart",AlxTopOnRewardVideoEvent.TAG)
        self.trackRewardedVideoAdVideoStart()
    }
    
    public func rewardVideoAdPlayEnd(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdPlayEnd",AlxTopOnRewardVideoEvent.TAG)
        self.trackRewardedVideoAdVideoEnd()
    }
    
    public func rewardVideoAdReward(_ ad: AlxRewardVideoAd) {
        NSLog("%@: rewardVideoAdReward",AlxTopOnRewardVideoEvent.TAG)
        self.isRewarded = true
        self.trackRewardedVideoAdRewarded()
    }
    
    public func rewardVideoAdPlayFail(_ ad: AlxRewardVideoAd, didFailWithError error: Error) {
        NSLog("%@: rewardVideoAdPlayFail",AlxTopOnRewardVideoEvent.TAG)
        self.trackRewardedVideoAdPlayWithError(error)
    }
    
    public override var networkUnitId: String {
        get {
            return self.serverInfo[AlxTopOnBaseManager.unitID] as? String ?? ""
        }
        set {
        }
    }
    
    deinit {
        NSLog("%@: deinit",AlxTopOnRewardVideoEvent.TAG)
    }
    
}
